ravenframework/SupervisedLearning/FeatureSelection/RFE.py

```diff
@@ -0,0 +1,774 @@
+    coefs = None
+    # number of subgroups
+    nGroups = max(len(self.subGroups), 1)
+    # features to select
+    nFeaturesToSelect = self.nFeaturesToSelect if self.nFeaturesToSelect is not None else self.maxNumberFeatures
+    # if both None ==> nFeatures/2
+    if nFeaturesToSelect is None:
+      nFeaturesToSelect = nFeatures // 2
+    else:
+      if nGroups > 1 and self.applyCrossCorrelation:
+        # we duouble the number of features because we apply the cross correlation later on to identify the
+        # number of features (nFeaturesToSelect)
+        nFeaturesToSelect *= 2
+    nFeaturesToSelect = nFeaturesToSelect if self.parametersToInclude is None else min(nFeaturesToSelect,nParams)
+
+    # get estimator parameter
+    originalParams = self.estimator.paramInput
+
+    # clustering appraoch here
+    if self.applyClusteringFiltering:
+      support_ = self.__applyClusteringPrefiltering(X, y, mask, support_)
+
+    # compute number of steps
+    setStep = int(self.step) if self.step > 1 else int(max(1, self.step * nParams))
+    # number of steps
+    nSteps = (int(np.sum(support_)) - nFeaturesToSelect)/setStep
+    # compute first step
+    diff = nSteps - int(nSteps)
+    firstStep = int(setStep * (1+diff))
+    step = firstStep
+    # we check the number of subgroups
+    outputspace = None
+    supportCandidates = []
+    outputSpaceToKeep = []
+    if nGroups > 1:
+      # re initialize support containers
+      groupFeaturesForRanking = copy.deepcopy(featuresForRanking)
+      groupRanking_ = copy.deepcopy(ranking_)
+      groupSupportOfSupport_ = copy.deepcopy(supportOfSupport_)
+
+    supportDataRFE = {'featuresForRanking':featuresForRanking,'mask':mask,'nFeatures':nFeatures,
+                   'nTargets':nTargets,'nParams':nParams,'targetsIds':targetsIds,
+                   'originalParams':originalParams,'supportOfSupport_':supportOfSupport_,
+                   'ranking_':ranking_,'nFeaturesToSelect':nFeaturesToSelect,'firstStep':step,
+                   'setStep':setStep,'subGroups':self.subGroups,
+                   'originalSupport':support_, 'parametersToInclude':self.parametersToInclude,
+                   'whichSpace':self.whichSpace}
+
+    if useParallel:
+      # send some data to workers
+      self.raiseADebug("Sending large data objects to Workers for parallel")
+      yRef = jhandler.sendDataToWorkers(y)
+      XRef = jhandler.sendDataToWorkers(X)
+      estimatorRef = jhandler.sendDataToWorkers(self.estimator)
+      self.raiseADebug("Large data objects have been sent to workers")
+      collectedOutput = 0
+      g = 0
+      while g < nGroups:
+        if jhandler.availability() > 0:
+          if nGroups > 1:
+            outputspace = self.subGroups[g]
+            self.raiseAMessage("Subgroupping with targets: {}".format(",".join(outputspace)))
+          else:
+            outputspace = None
+          prefix = f'subgroup_{g}'
+          if g > 0: supportDataRFE['firstStep'] = setStep
```

Comment:
Please put the "then" part on a separate line.

Reply:
done